### PR TITLE
check volumes for state when retrieving pool for configDrive creation

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

When querying for a storage pool to create a config drive on, only pools for those volumes that have the right state should be considered.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java [hunk 1]
```diff
@@ -322,27 +322,52 @@ public void commitMigration(NicProfile nic, Network network, VirtualMachineProfi
...
+            case Destroy:
+            case Destroying:
+            case UploadOp:
+            case Uploaded:
+            case NotUploaded:
+            case UploadInProgress:
+            case UploadError:
+            case UploadAbandoned:
+            default:
+                break;
```
Comment (rhtyd) [inline server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: All above cases can be removed?

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by PaulAngus** [comment]
Attempting to deploy a VM with a config drive on primary storage generates the following error:
(deployment works fine on secondary storage)

Asking ConfigDrive to release NicProfile[8-4-acef4b56-a86d-4519-affd-068869458240-null-null
2018-06-19 11:50:06,512 WARN  [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Unable to release some network resources.
java.lang.NullPointerException
  at com.cloud.network.element.ConfigDriveNetworkElement.deleteConfigDriveIso(ConfigDriveNetworkElement.java:423)



2018-06-19 11:50:06,007 DEBUG [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Deployment found  - P0=VM[User|i-2-4-VM], P0=Dest[Zone(Id)-Pod(Id)-Cluster(Id)-Host(Id)-Storage(Volume(Id|Type-->Pool(Id))] : Dest[Zone(1)-Pod(1)-Cluster(1)-Host(1)-Storage(Volume(6|ROOT-->Pool(2))]
2018-06-19 11:50:06,027 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) VM state transitted from :Starting to Starting with event: OperationRetryvm's original host id: null new host id: 1 host id before state transition: null
2018-06-19 11:50:06,033 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Hosts's actual total CPU: 6900 and CPU after applying overprovisioning: 13800
2018-06-19 11:50:06,033 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) We are allocating VM, increasing the used capacity of this host:1
2018-06-19 11:50:06,033 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Current Used CPU: 1500 , Free CPU:12300 ,Requested CPU: 500
2018-06-19 11:50:06,033 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Current Used RAM: 2147483648 , Free RAM:5368172544 ,Requested RAM: 536870912
2018-06-19 11:50:06,033 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) CPU STATS after allocation: for host: 1, old used: 1500, old reserved: 0, actual total: 6900, total with overprovisioning: 13800; new used:2000, reserved:0; requested cpu:500,alloc_from_last:false
2018-06-19 11:50:06,033 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) RAM STATS after allocation: for host: 1, old used: 2147483648, old reserved: 0, total: 7515656192; new used: 2684354560, reserved: 0; requested mem: 536870912,alloc_from_last:false
2018-06-19 11:50:06,040 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Network id=204 is already implemented
2018-06-19 11:50:06,100 DEBUG [c.c.n.NetworkModelImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Service SecurityGroup is not supported in the network id=204
2018-06-19 11:50:06,106 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Changing active number of nics for network id=204 on 1
2018-06-19 11:50:06,114 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Asking ConfigDrive to prepare for Nic[8-4-acef4b56-a86d-4519-affd-068869458240-null]
2018-06-19 11:50:06,131 DEBUG [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Cleaning up resources for the vm VM[User|i-2-4-VM] in Starting state
2018-06-19 11:50:06,137 DEBUG [c.c.a.t.Request] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Seq 1-5003499186008621072: Sending  { Cmd , MgmtId: 7020325439565, via: 1(pr2709-t2766-kvm-centos7-kvm1), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.StopCommand":{"isProxy":false,"checkBeforeCleanup":false,"forceStop":false,"volumesToDisconnect":[],"vmName":"i-2-4-VM","executeInSequence":false,"wait":0}}] }
2018-06-19 11:50:06,484 DEBUG [c.c.a.t.Request] (AgentManager-Handler-3:null) (logid:) Seq 1-5003499186008621072: Processing:  { Ans: , MgmtId: 7020325439565, via: 1, Ver: v1, Flags: 10, [{"com.cloud.agent.api.StopAnswer":{"result":true,"wait":0}}] }
2018-06-19 11:50:06,484 DEBUG [c.c.a.t.Request] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Seq 1-5003499186008621072: Received:  { Ans: , MgmtId: 7020325439565, via: 1(pr2709-t2766-kvm-centos7-kvm1), Ver: v1, Flags: 10, { StopAnswer } }
2018-06-19 11:50:06,493 DEBUG [c.c.n.NetworkModelImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Service SecurityGroup is not supported in the network id=204
2018-06-19 11:50:06,496 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Changing active number of nics for network id=204 on -1
2018-06-19 11:50:06,506 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Asking ConfigDrive to release NicProfile[8-4-acef4b56-a86d-4519-affd-068869458240-null-null
2018-06-19 11:50:06,512 WARN  [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Unable to release some network resources.
java.lang.NullPointerException
  at com.cloud.network.element.ConfigDriveNetworkElement.deleteConfigDriveIso(ConfigDriveNetworkElement.java:423)
  at com.cloud.network.element.ConfigDriveNetworkElement.release(ConfigDriveNetworkElement.java:147)
  at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.releaseNic(NetworkOrchestrator.java:1919)
  at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.release(NetworkOrchestrator.java:1845)
  at com.cloud.vm.VirtualMachineManagerImpl.cleanup(VirtualMachineManagerImpl.java:1567)
  at com.cloud.vm.VirtualMachineManagerImpl.orchestrateStart(VirtualMachineManagerImpl.java:1272)
  at com.cloud.vm.VirtualMachineManagerImpl.orchestrateStart(VirtualMachineManagerImpl.java:4930)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at com.cloud.vm.VmWorkJobHandlerProxy.handleVmWorkJob(VmWorkJobHandlerProxy.java:107)
  at com.cloud.vm.VirtualMachineManagerImpl.handleVmWorkJob(VirtualMachineManagerImpl.java:5093)
  at com.cloud.vm.VmWorkJobDispatcher.runJob(VmWorkJobDispatcher.java:102)
  at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:581)
  at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
  at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
  at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
  at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
  at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
  at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:529)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
2018-06-19 11:50:06,513 DEBUG [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) Successfully cleanued up resources for the vm VM[User|i-2-4-VM] in Starting state
2018-06-19 11:50:06,523 DEBUG [c.c.c.CapacityManagerImpl] (Work-Job-Executor-1:ctx-5b7ddba6 job-41/job-42 ctx-ef262bd0) (logid:1f612ec8) VM state transitted from :Starting to Stopped with event: OperationFailedvm's original host id: null new host id: null host id before state transition: 1

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java [hunk 2]
```diff
@@ -322,27 +322,52 @@ public void commitMigration(NicProfile nic, Network network, VirtualMachineProfi
...
-                }
-            }
-            if (dataStore == null) {
-                final List<VolumeVO> volumes = _volumeDao.findByInstanceAndType(profile.getVirtualMachine().getId(), Volume.Type.ROOT);
-                if (volumes != null && volumes.size() > 0) {
-                    dataStore = _dataStoreMgr.getDataStore(volumes.get(0).getPoolId(), DataStoreRole.Primary);
-                }
+            final List<VolumeVO> volumes = _volumeDao.findByInstanceAndType(profile.getVirtualMachine().getId(), Volume.Type.ROOT);
+            if (CollectionUtils.isNotEmpty(volumes)) {
+                dataStore = pickDataStoreFromVolumes(volumes);
```
Comment (rhtyd) [inline server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: We may need this code, it's because when VM is freshly deployed/started the `dest` can have storage disks which are not allocated but only in the destination plan; therefore we scan those planned disks to find the root disk and then find the datastore for the root disk. Note at this stage those planned disks may not have been persisted in the db.
If an existing (say stopped) VM is started, and destination plan may not have them then we simply query the volumes for the VM (which should be available from the db).

I think NPE is hit in case of freshly started VM because planned volume may not have been allocated and persisted in the db. IMO, this PR is not blocking RC and the refactoring PR can be targetted for 4.11.2.0.

**Comment by rhtyd** [comment]
@DaanHoogland I've tested and reviewed the PR, please see my notes in the diff section.

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: I agree that this is not blocking, but needs to be fixed as <quote>it is not taken into account that volumes may have a state of DESTROYED or EXPUNGING</quote> I'll be looking at a fix today.

Comment (rhtyd) [reply server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: okay @DaanHoogland, but if root volume is in destroyed or expunging that then VM is probably destroyed/expunged as well (probably corner case).

**Comment by DaanHoogland** [comment]
whether this works or not we can do with an integration test for it.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: @rhtyd I agree. I am really wondering if this is a red herring or a needed fix.

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java [hunk 3]
```diff
@@ -322,27 +324,78 @@ public void commitMigration(NicProfile nic, Network network, VirtualMachineProfi
...
+            if (volume.getVolumeType() == Volume.Type.ROOT) {
+                final StoragePool primaryPool = dest.getStorageForDisks().get(volume);
+                dataStore = _dataStoreMgr.getDataStore(primaryPool.getId(), DataStoreRole.Primary);
+                break;
+            }
+        }
+        return dataStore;
+    }
+
+    private DataStore savelyPickExistingRootVolumeDataStore(VirtualMachineProfile profile, DataStore dataStore) {
```
Comment (borisstoyanov) [inline server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: typo? 'safely'

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test
